Update logic for ShowMoreCard to determine when it needs to reset

bobsilverberg opened the pull request:
Fixes #10127

willdurand commented on src/amo/components/PermissionsCard/index.js:
```diff
@@ -57,6 +57,7 @@ export class PermissionsCardBase extends React.Component<Props> {
     return (
       <ShowMoreCard
         header={i18n.gettext('Permissions')}
+        childId={version.id}
```
`contentId` seems a bit more accurate (we can have more than a child)

willdurand commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -23,6 +23,7 @@ type UIStateType = {|
 |};
 
 type Props = {|
+  childId: number | string | null,
```
Could we enforce `string`?

bobsilverberg replied:
We could, but it's convenient for it to allow different values. Then the caller doesn't need to worry about conversion. This seems better to me.

willdurand commented on src/amo/components/ShowMoreCard/index.js:
```diff
@@ -98,9 +80,11 @@ export class ShowMoreCardBase extends React.Component<InternalProps> {
-    this.props.setUIState({
-      ...initialUIState,
-    });
+    if (this.props.uiState !== initialUIState) {
```
Is this still needed?

bobsilverberg replied:
I think so, in the case where the content isn't being truncated, but I can re-check.

bobsilverberg replied:
It doesn't seem to have an effect anymore, so it can be removed. I'll do so.

willdurand commented on src/amo/components/UserReview/index.js:
```diff
@@ -34,10 +34,12 @@ function reviewBody({
   content,
   html,
   id,
+  showMoreCardChildId = '',
```
mmm should we set it to `id` instead?

bobsilverberg replied:
Yeah, looking at it again, we don't really need `showMoreCardChildId` at all, and we can just always use `id`.

bobsilverberg commented:
@willdurand I addressed all of your comments but one, as I prefer to allow multiple types for `contentId` which makes it easier and more flexible for the callers.

willdurand commented on src/amo/components/UserReview/index.js:
```diff
@@ -76,8 +79,7 @@ export const UserReviewBase = (props: InternalProps) => {
 
-  const showMoreCardId =
-    review && review.id ? String(review.id) : 'loading-text';
+  const showMoreCardId = review && review.id ? String(review.id) : loadingId;
```
Interesting that you have to cast `review.id` to a `string`. I guess Flow doesn't like it but `contentId` is defined to accept both, so it should be possible without a cast. This is why I think it's better to restrict the `contentId` type: there is no choice to make, it's `'some-string'` or `String(id)`, and nothing else.

bobsilverberg replied:
Except for when it's null, which I think happens with `Addon`. Also, the reason it is being cast to a string there isn't because of `childId`, but because of the `ShowMoreCard`'s `id`.

My issue with changing `childId` to require a string is that then we'd have to cast in `Addon` and in `PermissionsCard`, and probably in anything in the future that uses it, and there's really no reason to require a string inside `ShowMoreCard`.

willdurand replied:
To be clear, my comment does not prevent you from landing this patch.

willdurand reviewed (approved):
r+wc
